flytekit/common/tasks/notebook.py

```diff
@@ -0,0 +1,542 @@
+        tm[k] = type_map[t]
+    return _packer(outputs, tm)
+
+
+def record_outputs(outputs=None):
+    p = convert_outputs(outputs)
+    return p.to_flyte_idl()
+
+# TODO: Support Client Mode
+def get_spark_context(spark_conf):
+    # We run in cluster-mode in Flyte.
+    # Ref https://github.com/lyft/flyteplugins/blob/master/go/tasks/v1/flytek8s/k8s_resource_adds.go#L46
+    if "FLYTE_INTERNAL_EXECUTION_ID" in _os.environ:
+        return SparkContext()
+
+    # Add system spark-conf for local/notebook based execution.
+    spark_conf.add(("spark.master", "local"))
+    conf = SparkConf().setAll(spark_conf)
+    return SparkContext(conf=conf)
+
+def python_notebook(
+        notebook_path='',
+        inputs=None,
+        outputs=None,
+        cache_version='',
+        retries=0,
+        deprecated='',
+        storage_request=None,
+        cpu_request=None,
+        gpu_request=None,
+        memory_request=None,
+        storage_limit=None,
+        cpu_limit=None,
+        gpu_limit=None,
+        memory_limit=None,
+        cache=False,
+        timeout=None,
+        environment=None,
+        cls=None,
+
+):
+    """
+    Decorator to create a Python Notebook Task definition.  This task will run as a single unit of work on the platform.
+
+    :rtype: flytekit.common.tasks.sdk_runnable.SdkNotebookTask
+    """
+    return SdkNotebookTask(
+            notebook_path=notebook_path,
+            inputs=inputs,
+            outputs=outputs,
+            task_type=_constants.SdkTaskType.PYTHON_TASK,
+            discovery_version=cache_version,
+            retries=retries,
+            deprecated=deprecated,
+            storage_request=storage_request,
+            cpu_request=cpu_request,
+            gpu_request=gpu_request,
+            memory_request=memory_request,
+            storage_limit=storage_limit,
+            cpu_limit=cpu_limit,
+            gpu_limit=gpu_limit,
+            memory_limit=memory_limit,
+            discoverable=cache,
+            timeout=timeout or _datetime.timedelta(seconds=0),
+            environment=environment,
+            custom={})
+
+
+class SdkNotebookTask(
+        _six.with_metaclass(sdk_bases.ExtendedSdkType, base_tasks.SdkTask)):
+
+    """
+    This class includes the additional logic for building a task that executes Notebooks.
+
+    """
+
+    def __init__(
+            self,
+            notebook_path,
+            inputs,
+            outputs,
+            task_type,
+            discovery_version,
+            retries,
+            deprecated,
+            storage_request,
+            cpu_request,
+            gpu_request,
+            memory_request,
+            storage_limit,
+            cpu_limit,
+            gpu_limit,
+            memory_limit,
+            discoverable,
+            timeout,
+            environment,
+            custom
+    ):
+
+        # Add output_notebook as an implicit output to the task.
+        outputs[OUTPUT_NOTEBOOK] = Types.Blob
```

Comment:
We should verify there is no name collision.

Also, lets make a TODO to introduce a Notebook type that derives from Blob so we can eventually know to render things nicely in the UI/etc.

Reply:
done